Dataset refactor

jindrahelcl opened the pull request:
This PR introduces a new way how to construct a dataset:

```ini
[train_data]
series=["source", "target", "source_prep", "computed"]
data=["/path/src", ("/path/tgt", <reader>), (<preprocessor>, "source"), <dataset_level_prep>]
outputs=[("target", "/path/out/tgt"), ("computed", "/path/out/cmp", <writer>)]
shuffled=True
buffer_size=1024
```

Details:
* Laziness of a dataset is now captured just in an object's field, not as a subclass
* In lazy dataset, the `buffer_size` determines how much of the data is being pre-fetched. In non-lazy dataset, this is always the size of the data.
* Shuffling can be done both on lazy and non-lazy datasets (lazy dataset will shuffle the buffer)
* `buffer_size` should be bigger than `batch_size` to avoid a warning.
* Writer objects are introduced. A special writer is `neuralmonkey.writers.auto.AutoWriter`, which preserves the functionality from the old codebase, which automatically selects the suitable writer given the type of the data being outputted.
* The old `load_dataset_from_files` function is re-written to work with new dataset to ensure backward compatibility, but a deprecation notice is logged whenever it is used.
* Small refactor was done regarding the sharing of the dataset batching. Now the batching is done exclusively in `learning_utils`, not in the TF manager's `execute` method. 
* Dataset series are now iterators, with the `dataset.get_series()` method returning a fresh iterator every time is done. Internally, the dataset stores factory functions which are called to create a new data iterator for each series.
* A tiny ugly aspect of this is the type resolution done on the `sources` argument. Originally, the `<preprocessor>` would follow the `"source"`  in the tuple, but this would clash with how the file readers are specified (`Tuple[str, Callable]`) so the series-level preprocessor had to become a `Tuple[Callable, str]`

varisd commented:
Why is the (<preprocessor>, "source") in this order and not ("source", <preprocessor>)?
I think that it would be clearer to have ('input_file/input_series', 'processor/reader')

jindrahelcl commented:
Because both reader and processor are `Callable`s and their types can overlap. (Readers are `Callable[[List[str]], Iterator[Any]]`, and preprocessors are `Callable[[List[Any], List[Any]]`, so there is overlap, for example for `Callable[[List[str]], List[str]]`.)

 When you want to recognize which is which according to its type (without having a parent class for `Reader` and `Preprocessor`), you need to switch the ordering, otherwise they are both `Tuple[str, Callable]`.

jindrahelcl commented:
... and you don't want to start inheriting these things, because it's not that simple because all the invariance/covariance and whatnot.

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,556 @@
+                raise ValueError("Lengths of data series do not match: {}"
+                                 .format(str(length_dict)))
+
+            self.length = next(iter(length_dict.values()))
```
Is the next(iter()) syntax necessary? Shouldn't length_dict.values()[0] be the same?

jindrahelcl replied:
```ipython
In [1]: a = {"barack": 67, "hussein": 12, "obama": 32}

In [2]: a.values()[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-53f4ff1337fa> in <module>()
----> 1 a.values()[0]

TypeError: 'dict_values' object does not support indexing
```

jindrahelcl replied:
https://stackoverflow.com/questions/3097866/access-an-arbitrary-element-in-a-dictionary-in-python

varisd replied:
Oh, now I see it.

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,556 @@
+        """Get the length of the dataset.
+
+        Returns:
+            The length of the dataset. If the dataset is lazy, returns None.
```
Documentation:
Actually, it throws an exception when lazy.

jindrahelcl replied:
good catch..

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,556 @@
+            name: The name of the series to fetch.
+
+        Returns:
+            The data series.
```
Documentation:
Isn't the function returning an data_series iterator? I would include it in the documentation if so (for clarity).

jindrahelcl replied:
Rephrased:

```
Returns:
    A freshly initialized iterator over the data series. 
```

varisd commented on neuralmonkey/writers/plain_text_writer.py~:
```diff
@@ -0,0 +1,31 @@
+from typing import TypeVar, Iterator, List, Any, Callable
```
Again, the .py~ file should probably not be included in the commit.

jindrahelcl replied:
dkuji

varisd commented on neuralmonkey/writers/numpy_writer.py~:
```diff
@@ -0,0 +1,37 @@
+from typing import Iterator, Dict, List
```
The .py~ file should be deleted

jindrahelcl replied:
děkuji

varisd commented on neuralmonkey/tests/test_dataset.py:
```diff
@@ -52,7 +53,8 @@ def test_glob(self):
-            self.assertEqual(dataset.get_series("data"), [["a"], ["b"], ["d"]])
+            series_iterator = dataset.get_series("data")
+            self.assertEqual(list(series_iterator), [["a"], ["b"], ["d"]])
 
```
It would probably be good to also include following tests:

- check whether the batching is done correctly (ideally with more than one series)
- check whether the shuffling does not break things (this might be tricky)
- check whether the buffering works

jindrahelcl replied:
hotovo

varisd commented on neuralmonkey/learning_utils.py:
```diff
@@ -390,18 +390,33 @@ def run_on_dataset(tf_manager: TensorFlowManager,
     result_data = {runner.output_series: result.outputs
                    for runner, result in zip(runners, all_results)}
 
+    # Run dataset-level preprocessing.
```
"postprocessing"

jindrahelcl replied:
dík

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,559 @@
+        while buf:
+            # Create the batch
+            name = "{}.batch.{}".format(self.name, batch_index)
+            rows = [buf.popleft() for _ in range(batch_size) if buf]
```
It should probably be mentioned in the documentation that it returns batches of size == buffer_size if buffer_size < batch_size

jindrahelcl replied:
to je bug, kterej nastane v extrémním případě. je tam warning, ale radši to nějak opravim, aby to vždycky dávalo batch_size

jindrahelcl replied:
Ono je vůbec otázka, jestli to chceme plnit a shufflovat po každý batchi, nebo jestli se tam nemá dát nějakej lower threshold pro velikost bufferu, pod kterej když to vleze, tak se to doplní na buffer_size

jindrahelcl replied:
Taky shufflování na deque je kvadratický, protože neni random access

jindrahelcl replied:
Navíc nevidim, proč jsem zvolil zrovna `deque`, když na tom volám jen `popleft`, `shuffle`, `extend`, `len` a testuju na prázdnotu

jindrahelcl replied:
ale jo, když se neshuffluje tak to tam chceš rvát jednou stranou a druhou to házet ven

jindrahelcl replied:
So, I propose:

1. Instead of a  warning, raise an exception, that the buffer is too small.
2. Document that in the docstring
3. Introduce another option `buffer_min_size`, that would be by default half the size of the buffer. (Think buffer_size of 1000 sentences, after 10 batches of 50 sentences each which we take from the buffer, the buffer gets refilled and reshuffled.)

jindrahelcl replied:
Anebo 1 škrtnout a přidat jen dokumentaci.

jindrahelcl replied:
htoovo

jlibovicky commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,559 @@
+            assert not self.lazy
+            buf = deque(zipped_iterator)
+        if self.shuffled:
+            random.shuffle(buf)  # type: ignore
```
Není efektivnější to lineárně převést na list, pak to lineárně zashuflovat a udělat z toho zase deque? Nebo je se to děje vevnitři v tom shufflu?

jindrahelcl replied:
Neděje. Viz co jsem napsal k Dušanovo review. Nejradši bych tam dal minimum size, pod kterou když to vleze, tak se to zashuffluje - klidně převedením na list a zpět. Ale nebude se to shufflovat po každý batchi

jindrahelcl replied:
snad hotovo. uvidime jestli projdou

jindrahelcl replied:
tests

jlibovicky commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,559 @@
+# pylint: disable=too-many-locals,too-many-branches
+def load(name: str,
+         series: List[str],
+         sources: List[SourceSpec],
```
Nelíbí se mi jméno `sources` protože koliduje se source a target v MT.

jindrahelcl replied:
A co se ti líbí?

jindrahelcl replied:
`data`?

jlibovicky replied:
Třeba nebo paths

jindrahelcl replied:
`paths` je blbý protože tam patří i series-level a dataset-level preprocesory. Proto se mi ani nelíbí data.

jindrahelcl replied:
Ale data docela jde

varisd replied:
dej tam series_source

jindrahelcl replied:
no, taky tam můžu dát `víčka_od_škraloupů`

jlibovicky replied:
Ale data mi pořád přijde lepší než sources.

jindrahelcl replied:
Ok, dám tam data

jindrahelcl replied:
hotovo

jindrahelcl commented:
I fixed the errors and removed the redundant `lazy` parameter. Now, everytime the `buffer_size` is specified, the dataset will behave lazily (předtim to spadlo, když lazy nebylo true a naopak když lazy bylo true a nebyl specifikovanej buffer size)

jindrahelcl commented:
Note that the data are actually stored in memory and are not re-read from the files, if the buffer_size is None. This is the expected behavior.

jlibovicky commented on neuralmonkey/config/parsing.py:
```diff
@@ -130,10 +130,12 @@ def _parse_list(string: str, vars_dict: VarsDict) -> List[Any]:
+    # types = [type(val) for val in values]
+
+    # if len(set(types)) > 1:
+    #     raise ParseError("List must of a same type, is: {}".format(types))
```
Proč to celý nesmažeš?

jindrahelcl replied:
Aby sis toho všimnul při review a kdyžtak řek že to nechcem rušit.

jlibovicky replied:
Od tý doby, co jedeme typeguardy, tak už to není potřeba, ne?

jindrahelcl replied:
řek bych že ne; smažu to

jlibovicky commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,572 @@
+SERIES_SOURCE = re.compile("s_([^_]*)$")
+SERIES_OUTPUT = re.compile("s_(.*)_out")
+
+
```
Tady bych možná přidal organizační komentář, že následují věci, co jenom nějak šupačí věci z konfiguráku.

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,572 @@
+
+    Keyword arguments:
+        name: The name of the dataset to use. If None (default), the name will
+              be inferred from the file names.
```
Tady to chce update docstringu. Z argumentu metody nevypliva, ze None je default + name se neinferuje s filenames, ale z nazvu objektu v configu, ne?

jindrahelcl replied:
na to kašli, to je deprecated funkce

varisd commented on neuralmonkey/dataset.py:
```diff
@@ -0,0 +1,572 @@
+        lazy: Boolean flag specifying whether to use lazy loading (useful for
+              large files). Note that the lazy dataset cannot be shuffled.
+              Defaults to False.
+        preprocessor: A callable used for preprocessing of the input sentences.
```
Tady se nemusi jednat pouze o vety, ne?

jindrahelcl replied:
na to kašli, to je deprecated funkce
